### ISSUE TITLE
AccessViolationException with SSL configuration (consumer)

### ISSUE THREAD

**Issue by themoah** [confluentinc/confluent-kafka-dotnet#125]
I'm running Confluent.Kafka client (0.9.4) with .Net framework 4.5.1. While it works fine without SSL, consumer client throws **AccessViolationException** with SSL configuration: 

```
{
            {"group.id", kafkaGroup}, 
            {"client.id", clientId}, 
            {"bootstrap.servers", kafkaServer}, 
            {"enable.auto.commit", true},
            {"security.protocol", "ssl" },
            {"ssl.key.location", "sslkey/client.keystore.jks"},
            { "ssl.key.password", "testtest" },
            { "auto.commit.interval.ms", 5000},
            { "default.topic.config", new Dictionary<string, object>()
                {
                    { "auto.offset.reset", "smallest" }
                }
            }

        }
```

**Comment by mhowlett** [comment]
looks like you need to review your configuration to me:
https://github.com/edenhill/librdkafka/wiki/Using-SSL-with-librdkafka

it's not good for us to be ever throwing an `AccessViolationException` though so I'll label this as bug and put it in the backlog.

please carefully review your config and update here if you are still seeing problems.

**Comment by themoah** [comment]
@mhowlett Double checked. If I comment "ssl.key.location" and run the code it work perfectly. If "ssl.key.location" is added code throws exception during consumer creation. 
Stacktrace (Changed some sensible data to ***)

```
at Confluent.Kafka.Impl.LibRdKafka.NativeMethods.rd_kafka_conf_destroy(IntPtr conf)
   at Confluent.Kafka.Impl.LibRdKafka.conf_destroy(IntPtr conf)
   at Confluent.Kafka.Impl.SafeKafkaHandle.Create(RdKafkaType type, IntPtr config)
   at Confluent.Kafka.Consumer..ctor(IEnumerable`1 config)
   at Confluent.Kafka.Consumer`2..ctor(IEnumerable`1 config, IDeserializer`1 keyDeserializer, IDeserializer`1 valueDeserializer)
   at ***.KafkaConsumer..ctor(ILogger logger, String kafkaServer, String kafkaGroup, String clientId, String topic, Action`1 callbackFunc) in ***\KafkaConsumer.cs:line 41
   at ***.KafkaConsumerTest.UnitTest_Connection() in ***\KafkaConsumerTest.cs:line 52
```

**Comment by edenhill** [comment]
Can you try specifying the full path to the keyt?

**Comment by themoah** [comment]
@edenhill same

**Comment by bjcryss** [comment]
Hi. I have (kind of) the same issue, but when I'm creating a publisher: 

Exception thrown: 'System.AccessViolationException' in Confluent.Kafka.dll)
InnerException=null.

I'm trying to connect to a cloudkarafka instance using SSL and the certificates they supply.
Any idea on what could be happening? Thanks!
(Sorry if i should have got this to another thread.)

**Comment by treziac** [comment]
Can you provide some context? (kafka version, ok, configuration, some code snippets...)

You may want to try upgrading librdkafka.redist to the latest prerelease (0.11.0-RC2) as some changes have been made on SSL (no need to upgrade confluent.Kafka, just add reference to librdkafka.redist package) The version is still prerelease and should not be used in prod yet, but you may verify if this solve your issue

**Comment by bjcryss** [comment]
Thank you for the recommending 0.11.0-RC2, but it did not fix the issue for me.

I have sent this issue also to cloudkarafka guys, but no response yet, so i'm still hoping to find a solution with you.

Some more details:
Cloudkarafka has given me the certificates in a single file called "env" with no extension, that is structured like this:
> CLOUDKARAFKA_CA="-----BEGIN CERTIFICATE-----[key]-----END CERTIFICATE-----"
CLOUDKARAFKA_CERT="-----BEGIN CERTIFICATE-----[key]-----END CERTIFICATE-----"
CLOUDKARAFKA_PRIVATE_KEY="-----BEGIN PRIVATE KEY-----[key]-----END PRIVATE KEY-----"

(where [key] is the long scrambled char string)

I copied the [key] values into their own specific files, and added them to the producer configuration (to ssl.ca.location, ssl.key.location, ssl.certificate.location) as specified on their documentation page(https://www.cloudkarafka.com/docs-nodejs.html - i used the node.js guide because they told me it is the most similar to .net in this case)

This configuration, leads me to the 'System.AccessViolationException' when instantiating the Producer. 
Stack trace: 

>    at Confluent.Kafka.Impl.LibRdKafka.NativeMethods.rd_kafka_conf_destroy(IntPtr conf)
   at Confluent.Kafka.Impl.LibRdKafka.conf_destroy(IntPtr conf)
   at Confluent.Kafka.Impl.SafeKafkaHandle.Create(RdKafkaType type, IntPtr config)
   at Confluent.Kafka.Producer..ctor(IEnumerable`1 config, Boolean manualPoll, Boolean disableDeliveryReports)
   at Confluent.Kafka.Producer`2..ctor(IEnumerable`1 config, ISerializer`1 keySerializer, ISerializer`1 valueSerializer, Boolean manualPoll, Boolean disableDeliveryReports)
   at Confluent.Kafka.Producer`2..ctor(IEnumerable`1 config, ISerializer`1 keySerializer, ISerializer`1 valueSerializer)
   at TestCloudKarafka.Program2.Main(String[] args) in D:\Tut\confluent-kafka-dotnet-master\TestCloudKarafka\Program2.cs:line 50
   at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
   at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

I need to specify now, that setting an empty string value instead of these certificates, i get "SSL handshake failed", but the Producer manages to get instantiated.

**Comment by bjcryss** [comment]
Issue resolved on my side, thank you for your quick involvement!
The issue was quite silly, for the separated certificates i saved into different files, by mistake i left the start&ending quotes inside the file. So the format of the pem was not quite correct. 
Thank you, keep on the good work!

**Comment by treziac** [comment]
Given the AccessException, it actually come from 
```
public static SafeKafkaHandle Create(RdKafkaType type, IntPtr config)
        {
            var errorStringBuilder = new StringBuilder(512);
            var skh = LibRdKafka.kafka_new(type, config, errorStringBuilder,
                    (UIntPtr) errorStringBuilder.Capacity);
            if (skh.IsInvalid)
            {
                LibRdKafka.conf_destroy(skh.DangerousGetHandle());
                throw new InvalidOperationException(errorStringBuilder.ToString());
            }
```

LibRdKafka.conf_destroy(skh.DangerousGetHandle()); throws (which seems normal as skh is invalid)
I'm not sure what we must do given conf in this case, rdkafka doc tate

> conf is an optional struct created with `rd_kafka_conf_new()` that will be used instead of the default configuration.
> The  conf object is freed by this function on success and must not be usedor destroyed by the application sub-sequently.

but when looking at rdkafka.c, it is also destroy when method fail if I'm not mistaking

@bjcryss @themoah fix on #263, you can take the nuget artefact here for test : https://ci.appveyor.com/project/mhowlett/confluent-kafka-dotnet/build/1.0.84
You should get an InvalidOperationException with proper error
(note that you wil have to target librdkafka.redist 0.11.0-RC2, RC1 had a bug)

**Comment by mhowlett** [comment]
closing as this was resolved, thanks @treziac !
